gossip/components/responder.c

```diff
@@ -6,219 +6,280 @@
-  }
-
-  if (trit_array_is_null(request->hash)) {
-    log_debug(RESPONDER_COMPONENT_LOGGER_ID,
-              "Responding to random tip request\n");
-    return random_tip_request(request);
-  }  // else
-  log_debug(RESPONDER_COMPONENT_LOGGER_ID,
-            "Responding to regular transaction request\n");
-  return regular_transaction_request(state, request, tx);
+  return ret;
 }
 
-static retcode_t reply_to_request(responder_state_t *const state,
-                                  transaction_request_t *const request,
-                                  iota_transaction_t const tx) {
+/**
+ * Responds to a request by:
+ * - sending a transaction to the requesting neighbor or
+ * - requesting a missing transaction
+ *
+ * @param responder The responder
+ * @param neighbor The requesting neighbor
+ * @param hash The request hash
+ * @param pack A stor pack containing a transaction
+ *
+ * @return a status code
+ */
+static retcode_t respond_to_request(responder_t const *const responder,
+                                    neighbor_t *const neighbor,
+                                    flex_trit_t const *const hash,
+                                    iota_stor_pack_t *const pack) {
   retcode_t ret = RC_OK;
 
-  if (state == NULL) {
-    return RC_RESPONDER_COMPONENT_NULL_STATE;
-  } else if (request == NULL) {
-    return RC_RESPONDER_COMPONENT_NULL_REQ;
-  } else if (request->neighbor == NULL) {
-    return RC_RESPONDER_COMPONENT_NULL_NEIGHBOR;
+  if (responder == NULL || neighbor == NULL || hash == NULL || pack == NULL) {
+    return RC_NULL_PARAM;
   }
 
-  if (tx != NULL) {
-    // Send transaction back to neighbor
-    flex_trit_t tx_trits[FLEX_TRIT_SIZE_8019];
+  if (pack->num_loaded != 0) {
+    // If a transaction or a random tip was found, sends it back to the neighbor
+    iota_transaction_t transaction = ((iota_transaction_t *)(pack->models))[0];
+    flex_trit_t transaction_flex_trits[FLEX_TRIT_SIZE_8019];
 
-    transaction_serialize_on_flex_trits(tx, tx_trits);
-    if ((ret = neighbor_send(state->node, request->neighbor, tx_trits)) !=
-        RC_OK) {
+    transaction_serialize_on_flex_trits(transaction, transaction_flex_trits);
+    if ((ret = neighbor_send(responder->node, neighbor,
+                             transaction_flex_trits)) != RC_OK) {
+      log_warning(RESPONDER_LOGGER_ID, "Sending transaction failed\n");
       return ret;
     }
+    // TODO Add transaction to cache
   } else {
-    // Transaction not found
-    // TODO(thibault): Randomly doesn't propagate request
-    if (trit_array_is_null(request->hash) == false) {
-      // Request is an actual missing transaction
-      if ((ret = request_transaction(&state->node->transaction_requester,
-                                     request->hash->trits, false)) != RC_OK) {
+    // If a transaction was requested but not found, requests it
+    if (!flex_trits_is_null(hash, FLEX_TRIT_SIZE_243) &&
+        ((double)rand() / (double)RAND_MAX) <
```

Comment:
Could we abstract rand() please? (so other impls can use a more secure (p)RNG or simialr)